ci/ray_ci/doc/load_doc_cache.py

```diff
@@ -0,0 +1,117 @@
+    latest_commit = subprocess.check_output(
+        [
+            "git",
+            "log",
+            "--pretty=format:%H",
+            "--first-parent",
+            "master",
+            "-n",
+            "1",
+        ]
+    ).decode("utf-8")
+    return latest_commit
+
+
+def fetch_cache_from_s3(commit, target_file_path):
+    """
+    Fetch doc cache archive from ray-ci-results S3 bucket
+
+    Args:
+        commit (str): The commit hash of the doc cache to fetch
+        target_file_path (str): The file path to save the doc cache archive
+    """
+    # Create an S3 client
+    s3 = boto3.client("s3")
+    s3_file_path = f"{DOC_BUILD_DIR_S3}/{commit}.tgz"
+    try:
+        print(f"Downloading {commit}.tgz from S3...")
+        s3.download_file(S3_BUCKET, s3_file_path, target_file_path)
+        print(f"Successfully downloaded {s3_file_path} to {target_file_path}")
+    except botocore.exceptions.ClientError as e:
+        print(f"Failed to download {s3_file_path} from S3: {str(e)}")
+        raise e
+
+
+def extract_cache(cache_path: str, doc_dir: str):
+    """
+    Extract the doc cache archive to overwrite the ray/doc directory
+
+    Args:
+        file_path (str): The file path of the doc cache archive
+    """
+    with tarfile.open(cache_path, "r:gz") as tar:
+        tar.extractall(doc_dir)
+    print(f"Extracted {cache_path} to {doc_dir}")
+
+
+def list_changed_and_added_files(ray_dir: str, latest_master_commit: str):
+    """
+    List all changed and added untracked files in the repo.
+    This is to prevent cache environment from updating timestamp of these files.
+    """
+    untracked_files = (
+        subprocess.check_output(
+            ["git", "ls-files", "--others"],
+            cwd=ray_dir,
+        )
+        .decode("utf-8")
+        .split(os.linesep)
+    )
+    modified_files = (
+        subprocess.check_output(
+            ["git", "ls-files", "--modified"],
+            cwd=ray_dir,
+        )
+        .decode("utf-8")
+        .split(os.linesep)
+    )
+    diff_files_with_master = (
+        subprocess.check_output(
+            ["git", "diff", "--name-only", latest_master_commit],
+            cwd=ray_dir,
+        )
+        .decode("utf-8")
+        .split(os.linesep)
+    )
+    filenames = []
+    for file in untracked_files + modified_files + diff_files_with_master:
+        filename = file.split(".")[0]  # Remove extension
+        if filename.startswith("doc/"):  # Remove "doc/" prefix
+            filename = filename.replace("doc/", "")
+        filenames.append(filename)
+    return filenames
+
+
+@click.command()
+@click.option("--ray-dir", default="/ray", help="Path to Ray repo")
+def main(ray_dir: str) -> None:
+    latest_master_commit = find_latest_master_commit()
+    # List all changed and added files in the repo
+    filenames = list_changed_and_added_files(ray_dir, latest_master_commit)
+    with open(
+        f"{ray_dir}/pending_files.txt", "w"
```

Comment:
nit: move `pending_files.txt` into a constant and reuse it in the other binary file